### DISCUSSION TITLE
PHPunit Tests failing after upgrading to Laravel 11

### DISCUSSION THREAD

**Discussion by javis** [laravel/framework#51652] [General] [answered]
After upgrading to Laravel 11 my tests are failing with a  **no such table** error for tables in a secondary DB connection. 

I am using in memory DBs with SQLITE for my tests.

Also, tests work fine when ran isolated, they only fail when running them all. So I think it is a problem related to the RefreshDatabase trait.

I read in the release notes for Laravel 11 that tests using memory dbs are faster now. Maybe this problem is related to this change?

**Comment by AidasK** [comment]
Have you found a cause of this? Facing the same issue

**Comment by EntroSolutions** [comment]
We are searching for fix too, everything is working perfect when I remove ":memory:" but when I put it back the tables are missing

**Comment by AidasK** [comment]
I found the reason why. For my some of the tests we were using sqlite and for some mysql and with laravel 11 you can not change databases as RefreshDatabase trait does not account for that. So to be able to test on both databases I have written these traits:

```php

namespace Tests\Traits;

use Illuminate\Contracts\Console\Kernel;
use Illuminate\Foundation\Testing\RefreshDatabase as BaseRefreshDatabase;
use Illuminate\Foundation\Testing\RefreshDatabaseState;
use Illuminate\Support\Facades\DB;
use Tests\TestCase;

trait RefreshDatabase
{
    use BaseRefreshDatabase;

    protected function beforeRefreshingDatabase()
    {
        DB::setDefaultConnection('sqlite-memory');
        config(['telescope.storage.database.connection' => 'sqlite-memory']);
        config(['queue.batching.database' => 'sqlite-memory']);
        config(['job-status.database_connection' => 'sqlite-memory']);
    }

    /**
     * Refresh a conventional test database.
     *
     * @return void
     */
    protected function refreshTestDatabase()
    {
        $connection = DB::getDefaultConnection();
        $migrated = TestCase::$migrated[$connection] ?? false;
        if (! $migrated) {
            $this->artisan('migrate:fresh', $this->migrateFreshUsing());

            $this->app[Kernel::class]->setArtisan(null);

            RefreshDatabaseState::$migrated = true;
            TestCase::$migrated[$connection] = true;
        }

        $this->beginDatabaseTransaction();
    }
}
```

And for mysql I use this:

```php

namespace Tests\Traits;

trait RefreshMysqlDatabase
{
    use RefreshDatabase;

    protected function beforeRefreshingDatabase()
    {
    }
}

```

You also need to update your databases config and add explicit `sqlite-memory` connection.

**Comment by javis** [comment]
Sorry for the delay, I haven't found an error yet. All my dbs for testing are SQLite in memory, so I'm not sure this will work for me. Anyways I'll give it a try, and let you know 

Thanks!

**Comment by javis** [comment, marked as answer]
I finally was able to find the fix.

In order for the new RefreshDatabase Trait to work with multiple database connections, you have to specify the names of the databases in your Test class using the `connectionsToTransact` property.

Example:

```
<?php

namespace Tests;


use Illuminate\Foundation\Testing\TestCase as BaseTestCase;
use Illuminate\Foundation\Testing\RefreshDatabase;

abstract class TestCase extends BaseTestCase
{
    use RefreshDatabase;

    protected $connectionsToTransact = [
        'sqlite',
        'legacy',
        'issues',
    ];
    
}
